packages/react-dom/src/events/DOMModernPluginEventSystem.js

```diff
@@ -305,3 +306,46 @@ export function attachElementListener(listener: ReactDOMListener): void {
 export function detachElementListener(listener: ReactDOMListener): void {
   // TODO
 }
+
+export function accumulateTwoPhaseListeners(event: ReactSyntheticEvent): void {
+  const phasedRegistrationNames = event.dispatchConfig.phasedRegistrationNames;
+  if (phasedRegistrationNames == null) {
+    return;
+  }
+  const {bubbled, captured} = phasedRegistrationNames;
+  const dispatchListeners = [];
+  const dispatchInstances = [];
+  let node = event._targetInst;
+  let hasListeners = false;
+
+  // Accumulate all instances and listeners via the target -> root path.
+  while (node !== null) {
+    // We only care for listeners that are on HostComponents (i.e. <div>)
+    if (node.tag === HostComponent) {
+      // Standard React on* listeners, i.e. onClick prop
+      const captureListener = getListener(node, captured);
+      if (captureListener != null) {
+        hasListeners = true;
+        // Capture listeners/instances should go at the start, so we
+        // unshift them to the start of the array.
+        dispatchListeners.unshift(captureListener);
+        dispatchInstances.unshift(node);
+      }
+      const bubbleListener = getListener(node, bubbled);
+      if (bubbleListener != null) {
+        hasListeners = true;
+        // Bubble listeners/instances should go at the end, so we
+        // push them to the end of the array.
+        dispatchListeners.push(bubbleListener);
+        dispatchInstances.push(node);
+      }
+    }
+    node = node.return;
+  }
+  // To prevent allocation to the event unless we actually
+  // have listeners we use the flag we would have set above.
+  if (hasListeners) {
```

Comment:
Wouldn’t checking the array length give us the same result?

Reply:
It would. Either work here. I can change to do that in a follow up.